crates/compiler/can/src/operator.rs

```diff
@@ -430,6 +479,88 @@ fn desugar_field<'a>(
     }
 }
 
+struct RecordBuilderArg<'a> {
+    closure: &'a Loc<Expr<'a>>,
+    apply_exprs: Vec<'a, &'a Loc<Expr<'a>>>,
+}
+
+fn record_builder_arg<'a>(
+    arena: &'a Bump,
+    region: Region,
+    fields: Collection<'a, Loc<RecordBuilderField<'a>>>,
+) -> RecordBuilderArg<'a> {
+    let mut record_fields = Vec::with_capacity_in(fields.len(), arena);
+    let mut apply_exprs = Vec::with_capacity_in(fields.len(), arena);
+    let mut apply_field_names = Vec::with_capacity_in(fields.len(), arena);
+
+    // Build the record that the closure will return and gather apply expressions
+
+    for field in fields.iter() {
+        let mut current = field.value;
+
+        let new_field = loop {
+            match current {
+                RecordBuilderField::Value(label, spaces, expr) => {
+                    break AssignedField::RequiredValue(label, spaces, expr)
+                }
+                RecordBuilderField::ApplyValue(label, _spaces, expr) => {
+                    apply_field_names.push(label);
+                    apply_exprs.push(expr);
+
+                    break AssignedField::LabelOnly(label);
+                }
+                RecordBuilderField::LabelOnly(label) => break AssignedField::LabelOnly(label),
+                RecordBuilderField::SpaceBefore(sub_field, _) => {
+                    current = *sub_field;
+                }
+                RecordBuilderField::SpaceAfter(sub_field, _) => {
+                    current = *sub_field;
+                }
+                RecordBuilderField::Malformed(malformed) => {
+                    break AssignedField::Malformed(malformed)
+                }
+            }
+        };
+
+        record_fields.push(Loc {
+            value: new_field,
+            region: field.region,
+        });
+    }
+
+    let record_fields = fields.replace_items(record_fields.into_bump_slice());
+
+    let mut body = arena.alloc(Loc {
+        value: Record(record_fields),
+        region,
+    });
+
+    // Construct the builder's closure
+    //
+    // { x, y, z: 3 }
+    // \y -> { x, y, z: 3 }
+    // \x -> \y -> { x, y, z: 3 }
+
+    for name in apply_field_names.iter().rev() {
+        let ident = roc_parse::ast::Pattern::Identifier(name.value);
+
+        let arg_pattern = arena.alloc(Loc {
+            value: ident,
+            region: name.region,
+        });
+
+        body = arena.alloc(Loc {
+            value: Closure(std::slice::from_ref(arg_pattern), body),
+            region,
+        });
+    }
```

Comment:
We don't need to handle it in this PR, but there is an edge case here we should be aware of: as it is, this transformation may introduce shadows. For example, suppose `x=1` is defined somewhere above the record builder, and the desugaring performs this transformation.

One way to address this is to prefix all lambdas with a unique symbol, for example transforming to `#x`. That will never introduce a shadowing problem. Another way is to relay to the canonicalizer that these transformations do not shadows from the surface syntax.

Reply:
Oh, great point! I will look into handling this tomorrow. I think we should address it in this PR.